src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs

```diff
@@ -0,0 +1,223 @@
+        }
+
+        [JsonRpcMethod(Methods.TextDocumentDocumentSymbolName)]
+        public async Task<object[]> GetTextDocumentDocumentSymbolsAsync(JToken input, CancellationToken cancellationToken)
+        {
+            var documentSymbolParams = input.ToObject<DocumentSymbolParams>();
+            return await this._protocol.GetDocumentSymbolsAsync(_workspace.CurrentSolution, documentSymbolParams, _clientCapabilities, cancellationToken).ConfigureAwait(false);
+        }
+
+        [JsonRpcMethod(Methods.TextDocumentFormattingName)]
+        public async Task<TextEdit[]> GetTextDocumentFormattingAsync(JToken input, CancellationToken cancellationToken)
+        {
+            var documentFormattingParams = input.ToObject<DocumentFormattingParams>();
+            return await this._protocol.FormatDocumentAsync(_workspace.CurrentSolution, documentFormattingParams, _clientCapabilities, cancellationToken).ConfigureAwait(false);
+        }
+
+        [JsonRpcMethod(Methods.TextDocumentOnTypeFormattingName)]
+        public async Task<TextEdit[]> GetTextDocumentFormattingOnTypeAsync(JToken input, CancellationToken cancellationToken)
+        {
+            var documentOnTypeFormattingParams = input.ToObject<DocumentOnTypeFormattingParams>();
+            return await this._protocol.FormatDocumentOnTypeAsync(_workspace.CurrentSolution, documentOnTypeFormattingParams, _clientCapabilities, cancellationToken).ConfigureAwait(false);
+        }
+
+        [JsonRpcMethod(Methods.TextDocumentImplementationName)]
+        public async Task<object> GetTextDocumentImplementationsAsync(JToken input, CancellationToken cancellationToken)
+        {
+            var textDocumentPositionParams = input.ToObject<TextDocumentPositionParams>();
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+            return await this._protocol.FindImplementationsAsync(_workspace.CurrentSolution, textDocumentPositionParams, _clientCapabilities, cancellationToken).ConfigureAwait(false);
+        }
+
+        [JsonRpcMethod(Methods.TextDocumentRangeFormattingName)]
+        public async Task<TextEdit[]> GetTextDocumentRangeFormattingAsync(JToken input, CancellationToken cancellationToken)
+        {
+            var documentRangeFormattingParams = input.ToObject<DocumentRangeFormattingParams>();
+            return await this._protocol.FormatDocumentRangeAsync(_workspace.CurrentSolution, documentRangeFormattingParams, _clientCapabilities, cancellationToken).ConfigureAwait(false);
+        }
+
+        [JsonRpcMethod(Methods.TextDocumentReferencesName)]
+        public async Task<object[]> GetTextDocumentReferencesAsync(JToken input, CancellationToken cancellationToken)
+        {
+            var referenceParams = input.ToObject<ReferenceParams>();
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
```

Comment:
out of curiousity, why do this?

Reply:
good question!  essentially, FAR calls to get third party references (xaml) and to get those references it expects to be called on the UI thread.  See https://github.com/dotnet/roslyn/blob/e0fbc0ef5475365f80edffcb71313e601aa3fbf2/src/EditorFeatures/Core/FindUsages/AbstractFindUsagesService.cs#L66

Reply:
definitely should comment this location as well :)

Reply:
Should the SwitchToMainThreadAsync() then be somewhere in LanguageServerProtocol.FindReferencesAsync() or basically right before the hop to the FindUsagesService? Because otherwise somebody could easy add a ConfigureAwait(false) in the middle of that and totally blow this up.

Reply:
(I'm presuming here that this part of the stack allows SwitchToMainThreadAsync() because either:

1. JTF rules apply, although I'm curious how that works through a StreamJsonRpc link, or
2. it's understood that no synchronous wait can be done on these at all, because StreamjsonRpc is not JTF-friendly.)

Reply:
Moved this back to the liveshare API for the moment until we can get rid of the UI dependency since it isn't safe to use jtf across stream json rpc.